### PR TITLE
Added macro variants with metric support matching the ones in CodeUtils

### PR DESCRIPTION
**Pull request by anush-apple**
- Added Verify/Return/Log/Succ/Exit XYZ WithMetric macro variants for use in metrics
- Added unit tests for these macro variants

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

File: src/tracing/metric_macros.h [hunk 1]
```diff
@@ -27,13 +27,258 @@
 #if MATTER_TRACING_ENABLED
 
 /**
- *  @def SuccessOrExitWithMetric(kMetriKey, error)
+ *  @def ReturnErrorOnFailureWithMetric(kMetricKey, expr)
+ *
+ *  @brief
+ *    This macros emits the specified metric with error code and returns the error code,
```
Comment (bzbarsky-apple) [inline src/tracing/metric_macros.h:33]: "This macro"

File: src/tracing/metric_macros.h [hunk 2]
```diff
@@ -27,13 +27,258 @@
...
+ *    other than CHIP_NO_ERROR. For an integer expression, this means non-zero.
+ *
+ *  Example usage:
+ *
+ *  @code
+ *    ReturnErrorOnFailureWithMetric(kMetricKey, channel->SendMsg(msg));
+ *  @endcode
+ *
+ *  @param[in]  kMetricKey  Metric key for the metric event to be emitted if the expr evaluates
+ *                          does not evaluate to CHIP_NO_ERROR. Value of the metric is to the
```
Comment (bzbarsky-apple) [inline src/tracing/metric_macros.h:44]: What does "evaluates does not evaluate" mean?

What does "is to the result" mean?

File: src/tracing/metric_macros.h [hunk 3]
```diff
@@ -27,13 +27,258 @@
...
+ *    error code as the value of the metric.
+ *
+ *  Example usage:
+ *
+ *  @code
+ *    ReturnLogErrorOnFailureWithMetric(kMetricKey, channel->SendMsg(msg));
+ *  @endcode
+ *
+ *  @param[in]  kMetricKey  Metric key for the metric event to be emitted if the expr evaluates
+ *                          does not evaluate to CHIP_NO_ERROR. Value of the metric is to the
```
Comment (bzbarsky-apple) [inline src/tracing/metric_macros.h:74]: Same questions as above.

File: src/tracing/metric_macros.h [hunk 4]
```diff
@@ -27,13 +27,258 @@
...
+ *    anything but CHIP_NO_ERROR, a metric with the specified key is emitted along with error as the value.
+ *
+ *  Example usage:
+ *
+ *  @code
+ *    ReturnOnFailureWithMetric(kMetricKey, channel->SendMsg(msg));
+ *  @endcode
+ *
+ *  @param[in]  kMetricKey  Metric key for the metric event to be emitted if the expr evaluates
+ *                          does not evaluate to CHIP_NO_ERROR. Value of the metric is to the
```
Comment (bzbarsky-apple) [inline src/tracing/metric_macros.h:105]: Same questions as above.

File: src/tracing/metric_macros.h [hunk 5]
```diff
@@ -27,13 +27,258 @@
...
+            return;                                                                                                                \
+        }                                                                                                                          \
+    } while (false)
+
+/**
+ *  @def VerifyOrReturnErrorWithMetric(kMetricKey, expr, code, ...)
+ *
+ *  @brief
+ *    Returns a specified error code if expression evaluates to false. If the expression evaluates
+ *    to false, a metric with the specified key is emitted with the value set to the code.
```
Comment (bzbarsky-apple) [inline src/tracing/metric_macros.h:154]: "to the error code"?  Or "to code"?  Either would be better than what we have here.

File: src/tracing/metric_macros.h [hunk 6]
```diff
@@ -27,13 +27,258 @@
...
+ *
+ *  Example usage:
+ *
+ *  @code
+ *    VerifyOrReturnErrorWithMetric(kMetricKey, param != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+ *  @endcode
+ *
+ *  @param[in]  kMetricKey  Metric key for the metric event to be emitted if the expr evaluates
+ *                          to false. Value of the metric is to code.
```
Comment (bzbarsky-apple) [inline src/tracing/metric_macros.h:163]: What does "to code" mean here?

File: src/tracing/metric_macros.h [hunk 7]
```diff
@@ -27,13 +27,258 @@
...
+ *    to false, a metric with the specified key is emitted with the value set to value.
+ *
+ *  Example usage:
+ *
+ *  @code
+ *    VerifyOrReturnValueWithMetric(kMetricKey, param != nullptr, Foo());
+ *  @endcode
+ *
+ *  @param[in]  kMetricKey  Metric key for the metric event to be emitted if the expr evaluates
+ *                          to false. Value of the metric is to value.
```
Comment (bzbarsky-apple) [inline src/tracing/metric_macros.h:185]: What does "to value" mean?

File: src/tracing/metric_macros.h [hunk 8]
```diff
@@ -27,13 +27,258 @@
...
+ *    with the value set to code.
+ *
+ *  Example usage:
+ *
+ *  @code
+ *    VerifyOrReturnLogErrorWithMetric(kMetricKey, param != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+ *  @endcode
+ *
+ *  @param[in]  kMetricKey  Metric key for the metric event to be emitted if the expr evaluates
+ *                          to false. Value of the metric is to code.
```
Comment (bzbarsky-apple) [inline src/tracing/metric_macros.h:216]: What does "to code" mean here?

File: src/tracing/metric_macros.h [hunk 9]
```diff
@@ -27,13 +27,258 @@
...
+ *
+ *  Example usage:
+ *
+ *  @code
+ *    ReturnErrorCodeWithMetricIf(kMetricKey, state == kInitialized, CHIP_NO_ERROR);
+ *    ReturnErrorCodeWithMetricIf(kMetricKey, state == kInitialized, CHIP_ERROR_INCORRECT_STATE);
+ *  @endcode
+ *
+ *  @param[in]  kMetricKey  Metric key for the metric event to be emitted if the expr evaluates
+ *                          to true. Value of the metric is to code.
```
Comment (bzbarsky-apple) [inline src/tracing/metric_macros.h:260]: "to code"?

**Pull request merged**